src/id_set.rs

```diff
@@ -0,0 +1,42 @@
+use std::collections::HashMap;
+
+/// A set that provides removal tokens when an item is added.
+#[derive(Debug, Clone)]
+pub(crate) struct IdSet<T> {
```

Comment:
It feels like there _must_ be an existing and more robust solution for this, but a quick crate search didn't turn anything up.